design/trait-work-as-patch-of-workload.md

```diff
@@ -0,0 +1,256 @@
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: NodeSelector
+            spec:
+              disktype: ssd
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: SecurityContext
+            spec:
+              runAsNonRoot: true
+```
+
+After the change of AppConfig deployed, NodeSelector and SecurityContext trait will work in a random order. Assuming the controller of NodeSelector trait
+will work first, then it will patch and update the Deployment.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      nodeSelector:
+        disktype: ssd
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+``` 
+
+This will cause first round re-create of pods, it won't trigger the canary trait to work, and the App will become unstable.
+
+Before the deployment become stable, the controller of SecurityContext trait will work and update the deployment again.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      securityContext:
+        runAsNonRoot: true
+      nodeSelector:
+        disktype: ssd
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+``` 
+
+This will cause a second round of re-create. Again, the canary trait won't work.
+
+There're no doubt too many risks here in this workflow. We can conclude into two problems:
+
+1. Pods can't be re-created several times, all changes of trait in one deploy can only affect pod once.
+2. Changes of trait which will affect the underlying workload must trigger new revision and make the canary trait work.
+
+In this proposal, we're going to fix these two probelms.
+
+## Goals
+
+1. make sure all changes of workload and trait only will affect underlying workload once.
+2. make sure underlying workload change will trigger revision create and trigger rollout/canary trait to work.
+
+## Proposal
+
+The overall idea of this proposal is to merge all traits and patch on the workload only once, and this action will
+trigger component revision generated automatically.
+
+To achieve this, we propose: 
+
+1. Add `patch` field into TraitDefinition, this will make the trait become a kind of patch trait.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: nodeselectors.extended.oam.dev
+spec:
+  patch: true 
+  appliesToWorkloads:
+    - core.oam.dev/v1alpha2.ContainerizedWorkload
+  definitionRef:
+    name: nodeselectors.extended.oam.dev
+```
+
+2. All patch traits MUST be created before workload emit, and oam-k8s-runtime will be pending until all patch traits are
```

Comment:
It's not clear about how to handle a trait then it patches workload CR during runtime (after app is deployed).